src/fields/mod.rs

```diff
@@ -172,6 +184,33 @@ pub trait LagrangePolynomialFieldElement: FieldElement {
     /// basis polynomial.
     // TODO: This could probably be a constant.
     fn sumcheck_p2_squared_minus_sumcheck_p2_mul_inv() -> Self;
+
+    /// The multiplicative inverse of this value.
+    fn mul_inv(&self) -> Self {
+        // We use Euler's theorem to compute inverses. There are more efficient algorithms which we
+        // will adopt in the future.
+        //
+        // https://en.wikipedia.org/wiki/Modular_multiplicative_inverse#Using_Euler's_theorem
+        self.pow(Self::modulus() - (BigUint::one() + BigUint::one()))
+    }
+
+    /// Raise a field element to some power.
+    fn pow(&self, mut exponent: BigUint) -> Self {
+        // Modular exponentiation from Schneier's _Applied Cryptography_, via Wikipedia
+        // https://en.wikipedia.org/wiki/Modular_exponentiation#Pseudocode
+        let mut out = Self::ONE;
+        let mut base = *self;
+
+        while exponent > BigUint::ZERO {
+            if exponent.is_odd() {
```

Comment:
this is a branch on secret data, right? and so would the exponent shift below?

Reply:
The exponent isn't secret in the use we make now, which is always `x ** MODULUS - 2`. But generally, yeah, we know we need to work on being constant time: https://github.com/abetterinternet/zk-cred-longfellow/issues/15